api/product.go

```diff
@@ -1,25 +1,116 @@
 package api
 
 import (
+	"errors"
+	"net/http"
+	"product-api/apperror"
+	"product-api/model"
+	"strconv"
+
 	"github.com/gin-gonic/gin"
 )
 
 func (h *APIHandler) GetListProduct(ctx *gin.Context) {
-	// TODO: handle get products
+	products, err := h.productService.GetList()
+	if err != nil {
+		ctx.JSON(http.StatusInternalServerError, model.NewErrorResponse(err.Error()))
+		return
+	}
+
+	ctx.JSON(http.StatusOK, products)
 }
 
 func (h *APIHandler) GetProductDetail(ctx *gin.Context) {
-	// TODO: handle get product by id
+	idParam, _ := ctx.Params.Get("id")
+	id, err := strconv.Atoi(idParam)
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, model.NewErrorResponse("invalid id param"))
+		return
+	}
+
+	pDetail, err := h.productService.GetByID(id)
+	if err != nil {
+		if errors.Is(err, apperror.ErrProductNotFound) {
+			ctx.JSON(http.StatusNotFound, model.NewErrorResponse(err.Error()))
+			return
+		}
+
+		ctx.JSON(http.StatusInternalServerError, model.NewErrorResponse(err.Error()))
+		return
+	}
+
+	ctx.JSON(http.StatusOK, pDetail)
 }
 
 func (h *APIHandler) StoreProduct(ctx *gin.Context) {
-	// TODO: handle saving product
+	var payload model.Product
+	err := ctx.ShouldBindJSON(&payload)
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, model.NewErrorResponse(err.Error()))
+		return
+	}
+
+	err = h.productService.Store(&payload)
+	if err != nil {
+		if errors.Is(err, apperror.ErrInvalidUserIdOrCategoryId) {
+			ctx.JSON(http.StatusBadRequest, model.NewErrorResponse(err.Error()))
+			return
+		}
+
+		ctx.JSON(http.StatusInternalServerError, model.NewErrorResponse(err.Error()))
+		return
+	}
+
+	ctx.JSON(http.StatusCreated, model.NewSuccessResponse("product added"))
 }
 
 func (h *APIHandler) DeleteProduct(ctx *gin.Context) {
-	// TODO: handle delete product
+	idParam, _ := ctx.Params.Get("id")
+	id, err := strconv.Atoi(idParam)
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, model.NewErrorResponse("invalid id param"))
+		return
+	}
+
+	err = h.productService.Delete(id)
+	if err != nil {
+		if errors.Is(err, apperror.ErrProductNotFound) {
+			ctx.JSON(http.StatusNotFound, model.NewErrorResponse(err.Error()))
+			return
+		}
+
+		ctx.JSON(http.StatusInternalServerError, model.NewErrorResponse(err.Error()))
+		return
+	}
+
+	ctx.JSON(http.StatusOK, model.NewSuccessResponse("product deleted"))
 }
 
 func (h *APIHandler) UpdateProduct(ctx *gin.Context) {
-	// TODO: handle update product
+	idParam, _ := ctx.Params.Get("id")
+	id, err := strconv.Atoi(idParam)
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, model.NewErrorResponse("invalid id param"))
+		return
+	}
+
+	var payload model.Product
```

Comment:
it's better to create a new DTO to store the updated payload, don't force to use model.Product, because not all fields are used during this update.

And please read the docs of the validator to implement validation for required fields, min length for the `name` fields, etc. 
ref: https://github.com/go-playground/validator